core/src/main/scala/com/criteo/cuttle/Workflow.scala

```diff
@@ -16,30 +17,145 @@ trait Workflow[S <: Scheduling] {
-      val edgesToRemove = edges.filter(_._2 == root)
-
-      edgesToRemove.foreach {
-        case edge @ (child, _, _) =>
-          edges.remove(edge)
-          if (!edges.exists(_._1 == child)) {
-            orderedRoots.add(child)
+  def roots: Set[Job[S]] = {
+    val childNodes = edges.map { case (child, _, _) => child }
+    vertices.filter(!childNodes.contains(_))
+  }
+
+  def leaves: Set[Job[S]] = {
+    val parentNodes = edges.map { case (_, parent, _) => parent }
+    vertices.filter(!parentNodes.contains(_))
+  }
+
+  // Returns a list of jobs in the workflow sorted topologically, using Kahn's algorithm. At the
+  // same time checks that there is no cycle.
+  private[cuttle] lazy val jobsInOrder: List[Job[S]] = topologicalSort((_, _) => ()) match {
+    case Some(sortedJobs) => sortedJobs
+    case None => throw new IllegalArgumentException("Workflow has at least one cycle")
+  }
+
+  /**
+    * @return the list of strongly connected components in the workflow with more than 1 node, which
+    * are the cycles in the graph
+    * @note Tarjan algorithm
+    */
+  def findCycles(): List[List[Job[S]]] = {
+    case class NodeWithMetadata(node: Job[S], children: mutable.Set[NodeWithMetadata] = mutable.Set.empty) {
+      var discoveryTime: Long = Long.MaxValue
+      // Discovery time of the earliest discovered ancestor
+      var earliestAncestor: Long = Long.MaxValue
+
+      def addChild(child: NodeWithMetadata): Boolean = children.add(child)
+
+      def removeChild(child: NodeWithMetadata): Boolean = children.remove(child)
+
+      def visited: Boolean = discoveryTime < Long.MaxValue
+
+      // Override hash computation to prevent stack overflow when trying to compute the hash of a node using the
+      // default hash method on workflows with cycles
+      override def hashCode(): Int = node.hashCode()
+    }
+
+    // Perform DFS on each non visited node and pop up elements each time a strongly connected component is identified
+    val stack = new mutable.Stack[NodeWithMetadata]
+    // Bookkeeping of nodes whose SCC has not yet been identified
+    val nodesInStack = mutable.Set.empty[NodeWithMetadata]
+    var time = 0L
+    val stronglyConnectedComponents = mutable.ListBuffer.empty[List[Job[S]]]
+
+    def dfs(node: NodeWithMetadata): Unit = {
+      stack.push(node)
+      nodesInStack += node
+      time += 1
+      node.discoveryTime = time
+      node.earliestAncestor = time
+      node.children.foreach { child =>
+        if (!child.visited) {
+          // (node -> child) is a tree edge
+          dfs(child)
+          node.earliestAncestor = Math.min(node.earliestAncestor, child.earliestAncestor)
+        } else {
+          if (nodesInStack.contains(child)) {
+            // (node -> child) is a back edge
+            node.earliestAncestor = Math.min(node.earliestAncestor, child.discoveryTime)
           }
+          // Otherwise (node -> child) is a cross edge and 'child' was already added to a SCC
```

Comment:
Actually it is not necessarily a cross edge. It could be a tree edge or a forward edge to a node already added to a SCC .